File: 0001~0500/0040. Combination Sum II/main.cpp
/** 
 * Source: https://leetcode.com/problems/combination-sum-ii/
 * Date: 2022/12/24
 * Skill: 
 * Runtime: 
 * Memory Usage: 
 * Time complexity:
 * Space complexity: 
 * Constraints: 
 *      1 <= candidates.length <= 100
 *      Each number in candidates may only be used once in the combination.
 *      The solution set must not contain duplicate combinations.
 */

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> res; 
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        
    }
    
    void backTracking(vector<int>& candidates, int target, int index, vector<int>& members) {
        
    }
};

static const auto io_sync_off = []() {
    // * turn off sync
    std::ios::sync_with_stdio(false);
    // * untie in/out streams
    std::cin.tie(nullptr);
    return nullptr;
}();

File: 0001~0500/0040. Combination Sum II/main.py
"""
Source: https://leetcode.com/problems/combination-sum-ii/description/
Date: 2023/1/1
Skill: Backtracking
Runtime: 44 ms, faster than 97.93%
Memory Usage: 14 MB, less than 59.86%
Time complexity: 
Space complexity: 
Constraints: 
    1 <= candidates[i] <= 50
"""

from typing import List

class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res, members, sz = [], [], len(candidates)

        def backTracking(start, remainder):
            if remainder == 0:
                res.append(list(members)) #! Important!
                return
            for i in range(start, sz):
                if candidates[i] > remainder: return #! Important!
                if i>start and candidates[i] == candidates[i-1]: continue
                members.append(candidates[i])
                backTracking(i+1, remainder-candidates[i])
                members.pop()

        candidates.sort()
        backTracking(0, target)
        return res

if __name__ == "__main__":
    s = Solution
    ans = s.combinationSum2(s, [2,5,2,1,2], 5)